File: apiEZPZ/src/main/java/com/example/apiezpz/auth/security/JwtTokenProvider.java
package com.example.apiezpz.auth.security;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.UnsupportedJwtException;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class JwtTokenProvider {
    private final JwtProperties jwtProperties;
    private static final Logger log = LoggerFactory.getLogger(JwtTokenProvider.class);

    public String generateAccessToken(String username) {
        Date now = new Date();
        Date expiryDate = new Date(now.getTime() +
                jwtProperties.getAccessTokenExpirationMinutes() * 60 * 1000);

        return Jwts.builder()
                .setSubject(username)
                .setIssuedAt(now)
                .setExpiration(expiryDate)
                .signWith(SignatureAlgorithm.HS256, jwtProperties.getSecret().getBytes())
                .compact();
    }

    // 리프레시 토큰 생성
    public String generateRefreshToken(String username) {
        Date now = new Date();
        Date expiryDate = new Date(now.getTime() +
                jwtProperties.getRefreshTokenExpirationMinutes() * 60 * 1000);

        return Jwts.builder()
                .setSubject(username)
                .setIssuedAt(now)
                .setExpiration(expiryDate)
                .signWith(SignatureAlgorithm.HS256, jwtProperties.getSecret().getBytes())
                .compact();
    }

    public long getRefreshTokenExpirationMinutes() {
        return jwtProperties.getRefreshTokenExpirationMinutes();
    }

    public String getUsernameFromToken(String token) {
        try {
            return Jwts.parserBuilder()
                    .setSigningKey(jwtProperties.getSecret().getBytes())
                    .build()
                    .parseClaimsJws(token)
                    .getBody()
                    .getSubject();
        } catch (ExpiredJwtException e) {
            return e.getClaims().getSubject();
        } catch (Exception e) {
            log.error("토큰에서 username 추출 중 오류 발생", e);
            return null;
        }
    }

    public LocalDateTime getTokenExpiryDateTime(String token) {
        Date expiration = Jwts.parserBuilder()
                .setSigningKey(jwtProperties.getSecret().getBytes())
                .build()
                .parseClaimsJws(token)
                .getBody()
                .getExpiration();
        
        return expiration.toInstant()
                .atZone(ZoneId.systemDefault())
                .toLocalDateTime();
    }

    // 토큰 유효성 검사
    public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder()
                .setSigningKey(jwtProperties.getSecret().getBytes())
                .build()
                .parseClaimsJws(token);
            return true;
        } catch (io.jsonwebtoken.security.SecurityException | MalformedJwtException e) {
            log.error("잘못된 JWT 서명입니다.");
        } catch (ExpiredJwtException e) {
            log.error("만료된 JWT 토큰입니다.");
        } catch (UnsupportedJwtException e) {
            log.error("지원되지 않는 JWT 토큰입니다.");
        } catch (IllegalArgumentException e) {
            log.error("JWT 토큰이 잘못되었습니다.");
        }
        return false;
    }

    // 토큰이 만료되었는지 확인하는 메서드 추가
    public boolean isTokenExpired(String token) {
        try {
            Jwts.parserBuilder()
                .setSigningKey(jwtProperties.getSecret().getBytes())
                .build()
                .parseClaimsJws(token);
            return false;
        } catch (ExpiredJwtException e) {
            return true;
        } catch (Exception e) {
            return true;
        }
    }
}

File: backend/src/main/kotlin/com/example/apiezpz/auth/security/JwtTokenProvider.kt
package com.example.apiezpz.auth.security

import com.example.apiezpz.auth.service.RedisCacheService
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtTokenProvider(
    private val jwtProperties: JwtProperties,
    private val redisCacheService: RedisCacheService
) {
    private val key: SecretKey = if (jwtProperties.secret.length >= 64) {
        Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray())
    } else {
        Jwts.SIG.HS512.key().build()
    }
    private val log = LoggerFactory.getLogger(javaClass)

    val accessTokenExpirationMinutes: Long
        get() = jwtProperties.accessTokenExpirationMinutes

    val refreshTokenExpirationMinutes: Long
        get() = jwtProperties.refreshTokenExpirationMinutes

    fun generateAccessToken(username: String): String {
        val accessToken = generateToken(username, accessTokenExpirationMinutes)
        redisCacheService.cacheAccessToken(username, accessToken, accessTokenExpirationMinutes)
        return accessToken
    }

    fun generateRefreshToken(username: String): String {
        val refreshToken = generateToken(username, refreshTokenExpirationMinutes)
        redisCacheService.cacheRefreshToken(username, refreshToken, refreshTokenExpirationMinutes)
        return refreshToken
    }

    private fun generateToken(username: String, expirationMinutes: Long): String {
        val now = Date()
        val expiryDate = Date(now.time + expirationMinutes * 60 * 1000)
        return Jwts.builder()
            .subject(username)
            .issuedAt(now)
            .expiration(expiryDate)
            .signWith(key)
            .compact()
    }

    fun validateToken(token: String): Boolean {
        try {
            Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(token)
            return true
        } catch (ex: Exception) {
            log.error("JWT 토큰 검증 실패: {}", ex.message)
            return false
        }
    }

    fun getUsernameFromToken(token: String): String {
        return getClaimsFromToken(token).subject
    }

    fun getTokenExpiryDateTime(token: String): java.time.LocalDateTime {
        val claims = getClaimsFromToken(token)
        val expirationDate = claims.expiration
        return expirationDate.toInstant()
            .atZone(java.time.ZoneId.systemDefault())
            .toLocalDateTime()
    }

    fun getRemainingTimeInMinutes(token: String): Long {
        val claims = getClaimsFromToken(token)
        val expirationTime = claims.expiration.time
        val currentTime = System.currentTimeMillis()
        val remainingTime = expirationTime - currentTime
        
        return if (remainingTime > 0) {
            remainingTime / (60 * 1000) // 밀리초를 분으로 변환
        } else {
            0
        }
    }

    fun getClaimsFromToken(token: String): Claims {
        return Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .payload
    }
} 